Update dependencies to 45b9f0a (rc2) (#269)
JoshOrndorff/recipes: text/3-entrees/fees.md

Substrate provides the
[`transaction_payment` pallet](https://crates.parity.io/pallet_transaction_payment/index.html) for
calculating and collecting fees for executing transactions. Fees are broken down into two
components:

-   Byte fee - A fee proportional to the transaction's length in bytes. The proportionality constant

>     is a parameter in the `transaction_payment` pallet.

Reviewer: Is it possible to link to some relevant attribute on the `transaction_payment` pallet here?

	pub const TransactionByteFee: u128 = 1;
}

impl transaction_payment::Trait for Runtime {
	type Currency = balances::Module<Runtime>;
	type OnTransactionPayment = ();
	type TransactionByteFee = TransactionByteFee;
	type WeightToFee = IdentityFee<Balance>;
	type FeeMultiplierUpdate = ();
}
```

## 1 to 1 Conversion

In many cases converting weight to fees one-to-one, as shown above, will suffice and can be
accomplished with
[`IdentityFee`](https://crates.parity.io/frame_support/weights/struct.IdentityFee.html). This
approach is also taken in the
[node template](https://github.com/paritytech/substrate/blob/2d39ec2c4aaec1cc0f91fcb91734de8f408dc1b2/bin/node-template/runtime/src/lib.rs#L246).
It is also possible to provide a type that makes a more complex calculation. Any type that
implements
[`WeightToFeePolynomial`](https://crates.parity.io/frame_support/weights/trait.WeightToFeePolynomial.html)
will suffice.

## Linear Conversion

Another common way to convert weight to fees is linearly. When converting linearly, the weight is
multiplied by a constant coefficient to determine the fee to charge. This is demonstrated in the
`weight-fee-runtime` with the `LinearWeightToFee` struct.

> We declare the struct with a associated type `C` which will provide the coefficient.

Reviewer: ```suggestion
We declare the struct with an associated type `C`, which will provide the coefficient.
```

```rust, ignore
pub struct LinearWeightToFee<C>(sp_std::marker::PhantomData<C>);
```

Then we implement `WeightToFeePolynomial` for it. When implementing this trait, your main job is to

> return set of

Reviewer: ```suggestion
return a set of
```

[`WeightToFeeCoefficient`](https://crates.parity.io/frame_support/weights/struct.WeightToFeeCoefficient.html)s.
These coefficients can have integer and fractional parts and be positive or negative. In our
`LinearWeightToFee` there is a single integer coefficient supplied by the associated type.

```rust, ignore
impl<C> WeightToFeePolynomial for LinearWeightToFee<C>
where
	C: Get<Balance>,
{
	type Balance = Balance;

	fn polynomial() -> WeightToFeeCoefficients<Self::Balance> {
		let coefficient = WeightToFeeCoefficient {
			coeff_integer: C::get(),
			coeff_frac: Perbill::zero(),
			negative: false,
			degree: 1,
		};

> 		// Return a smallvec of coefficients. Order does not need to match degrees

Reviewer: You have [a similar comment elsewhere in this PR](https://github.com/substrate-developer-hub/recipes/pull/269/files#diff-349b3dd3c5a6cfb0ad1b5850e089f95eR144), but it includes additional details. Do you want to update this one to match?

Author: Yeah I considered that too. Ultimately I decided it made more sense to introduce the consequences of negative coefficients in the `QuadraticWeightToFee` because that's where I actually use negative coefficients.

And if anyone wants to use `LinearWeightToFee` with a negative coefficient, they already have problems :)

		// because each coefficient has an explicit degree annotation.
		smallvec!(coefficient)
	}
}
```

This struct is reusable, and works with different coefficients. Using it looks like this.

```rust, ignore
parameter_types! {

> 	// Used with LinearWeightToFee conversion. Leaving this constant in tact when using other

Reviewer: ```suggestion
	// Used with LinearWeightToFee conversion. Leaving this constant intact when using other
```

Author: TIL 💡

	// conversion techniques is harmless.
	pub const FeeWeightRatio: u128 = 1_000;

	// --snip--
}

impl transaction_payment::Trait for Runtime {

	// Convert dispatch weight to a chargeable fee.
	type WeightToFee = LinearWeightToFee<FeeWeightRatio>;

	// --snip--
}
```

## Quadratic Conversion

More complex polynomials can also be used. When using complex polynomials, it is unlikely that your

> logic will be reused among multiple chains, so it is generally no worth the overhead of making the

Reviewer: ```suggestion
logic will be reused among multiple chains, so it is generally not worth the overhead of making the
```

> coefficients configurable. The `QuadraticWeightToFee` demonstrate a 2nd degree polynomial with

Reviewer: ```suggestion
coefficients configurable. The `QuadraticWeightToFee` demonstrates a 2nd-degree polynomial with
```
